File: Tests/ClausageTests/PromoScheduleTests.swift
import XCTest
@testable import Clausage

final class PromoScheduleTests: XCTestCase {

    private func utcDate(year: Int, month: Int, day: Int, hour: Int, minute: Int = 0, second: Int = 0) -> Date {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "UTC")!
        return cal.date(from: DateComponents(
            year: year, month: month, day: day,
            hour: hour, minute: minute, second: second
        ))!
    }

    // MARK: - currentStatus

    func testBeforePromoIsNotStarted() {
        let date = utcDate(year: 2026, month: 3, day: 12, hour: 23, minute: 59)
        XCTAssertEqual(PromoSchedule.currentStatus(at: date), .notStarted)
    }

    func testAfterPromoIsEnded() {
        let date = utcDate(year: 2026, month: 3, day: 28, hour: 7, minute: 0)
        XCTAssertEqual(PromoSchedule.currentStatus(at: date), .ended)
    }

    func testWeekdayPeakHoursIsPeak1x() {
        // Wednesday March 18, 2026 at 14:00 UTC (weekday, within 12-18 UTC)
        let date = utcDate(year: 2026, month: 3, day: 18, hour: 14)
        XCTAssertEqual(PromoSchedule.currentStatus(at: date), .peak1x)
    }

    func testWeekdayOffPeakIsActive2x() {
        // Wednesday March 18, 2026 at 08:00 UTC (weekday, before peak)
        let date = utcDate(year: 2026, month: 3, day: 18, hour: 8)
        XCTAssertEqual(PromoSchedule.currentStatus(at: date), .active2x)
    }

    func testWeekdayAfterPeakIsActive2x() {
        // Wednesday March 18, 2026 at 19:00 UTC (weekday, after peak)
        let date = utcDate(year: 2026, month: 3, day: 18, hour: 19)
        XCTAssertEqual(PromoSchedule.currentStatus(at: date), .active2x)
    }

    func testWeekendIsAlways2x() {
        // Saturday March 14, 2026 at 14:00 UTC (weekend, during what would be peak)
        let date = utcDate(year: 2026, month: 3, day: 14, hour: 14)
        XCTAssertEqual(PromoSchedule.currentStatus(at: date), .active2x)
    }

    func testSundayIsAlways2x() {
        let date = utcDate(year: 2026, month: 3, day: 15, hour: 14)
        XCTAssertEqual(PromoSchedule.currentStatus(at: date), .active2x)
    }

    func testPeakBoundaryStart() {
        // Exactly at 12:00 UTC on a weekday = peak
        let date = utcDate(year: 2026, month: 3, day: 18, hour: 12, minute: 0, second: 0)
        XCTAssertEqual(PromoSchedule.currentStatus(at: date), .peak1x)
    }

    func testPeakBoundaryEnd() {
        // Exactly at 18:00 UTC on a weekday = off-peak (>= 18 is off)
        let date = utcDate(year: 2026, month: 3, day: 18, hour: 18, minute: 0, second: 0)
        XCTAssertEqual(PromoSchedule.currentStatus(at: date), .active2x)
    }

    func testPromoStartBoundary() {
        let date = utcDate(year: 2026, month: 3, day: 13, hour: 0, minute: 0, second: 0)
        // March 13 is a Friday, 00:00 UTC = off-peak
        XCTAssertEqual(PromoSchedule.currentStatus(at: date), .active2x)
    }

    // MARK: - nextTransition

    func testNextTransitionFromNotStarted() {
        let date = utcDate(year: 2026, month: 3, day: 12, hour: 12)
        let transition = PromoSchedule.nextTransition(from: date)
        XCTAssertNotNil(transition)
        XCTAssertEqual(transition?.nextStatus, .active2x)
        XCTAssertEqual(transition?.date, PromoSchedule.promoStart)
    }

    func testNextTransitionFromEnded() {
        let date = utcDate(year: 2026, month: 3, day: 29, hour: 12)
        let transition = PromoSchedule.nextTransition(from: date)
        XCTAssertNil(transition)
    }

    func testNextTransitionFromPeakGoesTo2x() {
        // During peak, next transition is end of peak (18:00 UTC)
        let date = utcDate(year: 2026, month: 3, day: 18, hour: 14)
        let transition = PromoSchedule.nextTransition(from: date)
        XCTAssertNotNil(transition)
        XCTAssertEqual(transition?.nextStatus, .active2x)
    }

    func testNextTransitionFrom2xBeforePeakGoesToPeak() {
        // Before peak on a weekday, next transition is peak start
        let date = utcDate(year: 2026, month: 3, day: 18, hour: 8)
        let transition = PromoSchedule.nextTransition(from: date)
        XCTAssertNotNil(transition)
        XCTAssertEqual(transition?.nextStatus, .peak1x)
    }

    func testNextTransitionFromWeekendGoesToMondayPeak() {
        // Saturday, next transition is Monday 12:00 UTC
        let date = utcDate(year: 2026, month: 3, day: 14, hour: 14)
        let transition = PromoSchedule.nextTransition(from: date)
        XCTAssertNotNil(transition)
        XCTAssertEqual(transition?.nextStatus, .peak1x)

        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "UTC")!
        let weekday = cal.component(.weekday, from: transition!.date)
        XCTAssertEqual(weekday, 2, "Should transition to Monday")
    }

    // MARK: - Display strings

    func testPeakHoursLocalStringNotEmpty() {
        let str = PromoSchedule.peakHoursLocalString()
        XCTAssertFalse(str.isEmpty)
        XCTAssertTrue(str.contains("-"), "Should contain a dash separator")
    }

    func testPromoEndLocalStringNotEmpty() {
        let str = PromoSchedule.promoEndLocalString()
        XCTAssertFalse(str.isEmpty)
    }
}

File: windows/Tests/PromoScheduleTests.cs
using Clausage.Models;
using Xunit;

namespace Clausage.Tests;

public class PromoScheduleTests
{
    private static DateTime Utc(int year, int month, int day, int hour = 0, int minute = 0) =>
        new(year, month, day, hour, minute, 0, DateTimeKind.Utc);

    private static PromoSchedule MakeSchedule(
        string start = "2026-03-13T00:00:00Z",
        string end = "2026-03-28T06:59:59Z",
        int peakStart = 12, int peakEnd = 18)
    {
        var schedule = new PromoSchedule();
        schedule.Update(new PromoConfig(true, start, end, peakStart, peakEnd, "test"));
        return schedule;
    }

    [Fact]
    public void Disabled_When_No_Config()
    {
        var schedule = new PromoSchedule();
        Assert.Equal(PromoStatus.Disabled, schedule.CurrentStatus());
    }

    [Fact]
    public void Disabled_When_Config_Disabled()
    {
        var schedule = new PromoSchedule();
        schedule.Update(new PromoConfig(false, "", "", 12, 18, ""));
        Assert.Equal(PromoStatus.Disabled, schedule.CurrentStatus());
    }

    [Fact]
    public void Not_Started()
    {
        var schedule = MakeSchedule();
        Assert.Equal(PromoStatus.NotStarted, schedule.CurrentStatus(Utc(2026, 3, 12, 23)));
    }

    [Fact]
    public void Ended()
    {
        var schedule = MakeSchedule();
        Assert.Equal(PromoStatus.Ended, schedule.CurrentStatus(Utc(2026, 3, 29)));
    }

    [Fact]
    public void Active2x_OffPeak_Weekday()
    {
        var schedule = MakeSchedule();
        // Monday 2026-03-16, 8 AM UTC
        Assert.Equal(PromoStatus.Active2x, schedule.CurrentStatus(Utc(2026, 3, 16, 8)));
    }

    [Fact]
    public void Peak1x_Weekday()
    {
        var schedule = MakeSchedule();
        // Monday 2026-03-16, 14:00 UTC
        Assert.Equal(PromoStatus.Peak1x, schedule.CurrentStatus(Utc(2026, 3, 16, 14)));
    }

    [Fact]
    public void Active2x_Evening_Weekday()
    {
        var schedule = MakeSchedule();
        // Monday 2026-03-16, 20:00 UTC
        Assert.Equal(PromoStatus.Active2x, schedule.CurrentStatus(Utc(2026, 3, 16, 20)));
    }

    [Fact]
    public void Active2x_Weekend()
    {
        var schedule = MakeSchedule();
        // Saturday 2026-03-14, 14:00 UTC (peak hour but weekend)
        Assert.Equal(PromoStatus.Active2x, schedule.CurrentStatus(Utc(2026, 3, 14, 14)));
        // Sunday 2026-03-15, 10:00 UTC
        Assert.Equal(PromoStatus.Active2x, schedule.CurrentStatus(Utc(2026, 3, 15, 10)));
    }

    [Fact]
    public void NextTransition_From_NotStarted()
    {
        var schedule = MakeSchedule();
        var t = schedule.NextTransition(Utc(2026, 3, 12, 23));
        Assert.NotNull(t);
        Assert.Equal(PromoStatus.Active2x, t!.Value.NextStatus);
    }

    [Fact]
    public void NextTransition_From_2x_To_Peak()
    {
        var schedule = MakeSchedule();
        // Monday morning -> peak at 12:00
        var t = schedule.NextTransition(Utc(2026, 3, 16, 8));
        Assert.NotNull(t);
        Assert.Equal(PromoStatus.Peak1x, t!.Value.NextStatus);
        Assert.Equal(12, t.Value.Date.Hour);
    }

    [Fact]
    public void NextTransition_From_Peak_To_2x()
    {
        var schedule = MakeSchedule();
        // During peak -> 2x at 18:00
        var t = schedule.NextTransition(Utc(2026, 3, 16, 14));
        Assert.NotNull(t);
        Assert.Equal(PromoStatus.Active2x, t!.Value.NextStatus);
        Assert.Equal(18, t.Value.Date.Hour);
    }

    [Fact]
    public void NextTransition_From_Weekend_To_Monday_Peak()
    {
        var schedule = MakeSchedule();
        var t = schedule.NextTransition(Utc(2026, 3, 14, 10));
        Assert.NotNull(t);
        Assert.Equal(PromoStatus.Peak1x, t!.Value.NextStatus);
        Assert.Equal(DayOfWeek.Monday, t.Value.Date.DayOfWeek);
    }

    [Fact]
    public void NextTransition_Ended()
    {
        var schedule = MakeSchedule();
        Assert.Null(schedule.NextTransition(Utc(2026, 3, 29)));
    }

    [Fact]
    public void PeakHoursLocalString_Contains_Dash()
    {
        var schedule = MakeSchedule();
        Assert.Contains("-", schedule.PeakHoursLocalString());
    }

    [Fact]
    public void PromoEndLocalString_Contains_Year()
    {
        var schedule = MakeSchedule();
        Assert.Contains("2026", schedule.PromoEndLocalString());
    }
}
